Added new command for creating new databases

sschlein opened the pull request:
I find myself creating new databases in Sequel Pro each time I start a new project. Unfortunately, this is the only reason why I mostly open Sequel Pro and so I've added this as a valet command:

valet db new_db_name

I think it is a good fit for valet because the local mysql/mariadb user is root with no password and will be accessible via valet easily.

jerguslejko commented:
I would personally not go for this for couple reasons:

- valet is a web server - not concerned about a database at all
- what if someone uses different username or password? Or, what if they use mariadb/postgres/etc?

sschlein commented:
Hi @jerguslejko 

in my opinion, valet is not a web server - valet is a development environment. This allows it to bundle additional functionality with it and creating a new database for a new project is a common use case.

It also works with mariadb :smile: and I'm sure it is a good solution for most people developing with valet locally. It not a replacement for Sequel Pro or other database management apps but it just makes your day more convenient - that is what Laravel is about.

jerguslejko commented:
Alright, I personally use a password for my local mysql installation. What then?

(I did not know that `mysql -uroot ...` also works for _mariadb_ - that's cool.)

Anyway, if _valet_ provides a command for creating a database, shouldn't it provide one for deleting it? Also, valet is primarily designed for laravel, so it definitely should work with all supported database drivers.

No. What I'm trying to say is that this tool is not designed for this task. And this feature seems rather _"half baked"_.

sschlein commented:
Honestly... valet is the right tool for 80% of standard web development projects and the documentation comes with a local MariaDB installation without a password. The process of creating new databases is also much more common than to delete old ones and creation is required for each new project.

It's not about Laravel, it's about your local development environment.

Personally, I don't see the need for adding additional complexity which solves all edge cases 😺

drbyte commented:
I'm on the fence: I don't feel like I "need" this in Valet, but if it were there I "might" use it. (It'd be faster than my current workflow, and definitely less typing even from CLI.)

Either way, if one simply takes the proposed code and puts it into a new .php file in their `~/.valet/extensions` directory (ie: `make_db.php` or whatever), the command is instantly available to them if Valet is installed. (All files in the `extensions` folder are loaded after all the default commands are set up.)

I'm gonna do exactly that. Although I'm gonna change the command from `valet db` to `valet make:db` just cuz I like the expressiveness of including `make:` much like `artisan` does. That's just my preference though.

sschlein commented:
For the time beeing, I also removed my custom valet locally and added the extension like you've described it. I wasn't aware that this is possible because there is no hint in the docs.. even if the code to load valet extension is directly behind my changes ;-)

Thanks @drbyte !

antonioribeiro commented:
I like the idea, but my main concern about this is it being MySQL only. I rarely use it. Yeah, PostgreSQL has those nice `createdb` and `dropdb` commands, but still, if you are creating a command for something that is supposed to work to any developer in the world, you have to cover, at least, the range of database servers the Laravel Framework covers, or you don't create it at all. IMHO.

sepehr commented:
I don't think that's something Valet should be responsible for. What's the problem with a shell alias or helper function?

sschlein commented:
@sepehr do you have a shell alias or helper function configured in your terminal? I bet: No. Do you have valet? I guess: Yes.

With the feature, you would be able to create a new database within seconds without additional configuration or setup 😃 

Laravel and all tools around are about convenience.

sepehr commented:
@sschlein I agree with you that it's all about convenience. Setting up such lightweight dev env with features like https, sharing, etc. was hard, and more importantly, not easily repeatable before Valet.

But IMHO, adding a database creation feature is not one of those features that brings convenience, at least for me. It's such an easy work.

And yes, I do have a database creation helper, plus a few others:

```shell
# checks if command exists with zero status code
exists() {
    command -v $1 &> /dev/null
}

# create a new mysql database
mkdb() {
	if ! exists mysql; then
		echo 'MySQL is not available.'
		return 1
	fi

	if [ -z "$1" ]; then
		echo 'No database name provided.'
		return 1
	fi

	mysql -u root -p -e "drop database if exists $1; create database $1 default character set utf8 default collate utf8_general_ci;"
}

# hub wrapper
# requires piotrmurach/github_cli
git() {
    if [ "$1" != "delete" ]; then
        hub $@
        return
    fi

    if [ -z "$2" ]; then
        echo "No repo name given."
        return 1
    fi

    if ! exists gcli; then
        echo "gcli is not available. run: gem install github_cli"
        return 1
    fi

    GCLI_USER=$(gcli config user.login)
    if [ -z "$GCLI_USER" ]; then
        echo "gcli's user.login config is not set. see: gcli help config"
        return 1
    fi

    gcli repo delete $GCLI_USER $2
}

# todolist wrapper
t() {
    if [[ -f "./.todos.json" || "$1" == "init" ]]; then
        todolist $@
        return
    fi

    WDIR=`pwd`
    cd $HOME
	
	todolist $@
	
	cd $WDIR
}

# vagrant wrapper for homestead
homestead() {
    ( cd ~/Homestead && vagrant $* )
}

# fixes file/dir permissions
fixperms() {
    if [ -z "$1" ]; then
        echo "Please specify the path to directory."
        return
    fi

    confirm "Are sure that you want to fix permissions on \"$1\"? [y/N]" || return

    # chmod all files to 644
    # chmod all dires to 755
    echo ""
    echo "== Fixing file permissions..."
    echo "`find $1 -type f -print -exec chmod u=rw,g=r,o=r {} \; | wc -l` files processed."
    echo "`find $1 -type d -print -exec chmod u=rwx,g=rx,o=rx {} \; | wc -l` directories processed."

    # Remove "com.apple.quarantine" flag (OSX)
    echo ""
    echo "== Removing files from OSX quarantine, if any..."
    find $1 -exec xattr -d com.apple.quarantine {} 2&>1 /dev/null \;
    echo "    Done!"

    # Unlock (OSX)
    echo ""
    echo "== Unlocking locked files, if any..."
    echo "    Done!"
    chflags -R nouchg $1
    chflags -R noschg $1
}

# makes the file/dir apache writeable
a2mkw() {
    WWW_USER=`ps aux | egrep 'apache|httpd' | awk '{ print $1 }' | sed '1 d' | sort | uniq | awk '{print}' ORS=' ' | sed "s/\b$(whoami)\|root\b//g" | xargs`

    echo ""
    echo "== Setting up writables..."
    echo "= Webserver is running by \"$WWW_USER\""

    chgrp -v -R $WWW_USER $1
    chmod -v -R u=rwx,g=rwx,o=rx $1
    echo ""
}
```

adamwathan commented:
I don't dislike the idea but going to hold off on it for now I think. Can definitely use that extensions directory in the mean time 👍

sschlein commented:
For everyone finding this PR, please see this gist for the extension:

https://gist.github.com/sschlein/a5c17edd1a2a7f8a7d4dc3cf48b928aa
